### ISSUE TITLE
LoadIdentityClut() and ATIRadeonPerf

### ISSUE THREAD

**Issue by qx1147** [Psychtoolbox-3/Psychtoolbox-3#402]
First of all, I wonder how the LUT formulas in LoadIdentityClut() have been found and how the graphics output was tested.
Anyway, I just discovered that AMD has changed, in 2014 already, the interpretation of the dither state parameter for ADL_Display_DitherState_Get/Set() in an incompatible way. This basically makes the ATIRadeonPerf utility enable dithering when it is supposed to disable it, if graphics drivers released later than - I guess - mid 2014 are installed. It affects Windows but probably also Linux. I am using Windows7 64bit and Epiphan's DVI2USB3.0 frame grabber (via a passive DVI splitter) for checking the graphics card output. 

The parameter values used for ADL_Display_DitherState_Get/Set() can be found in the adl_defines.h file of the ADL_SDK:

ADL_SDK 6 (released late 2013):
```
#define ADL_DL_DISPLAY_DITHER_UNKNOWN    0
#define ADL_DL_DISPLAY_DITHER_DISABLED   1
#define ADL_DL_DISPLAY_DITHER_ENABLED    2
```

ADL_SDK 7 (released mid 2014) and later:
```
/// Dithering disabled.
#define ADL_DL_DISPLAY_DITHER_DISABLED              0
/// Use default driver settings for dithering. Note that the default setting could be dithering disabled.
#define ADL_DL_DISPLAY_DITHER_DRIVER_DEFAULT        1
/// Temporal dithering to 6 bpc. Note that if the input is 12 bits, the two least significant bits will be truncated.
#define ADL_DL_DISPLAY_DITHER_FM6                   2
/// Temporal dithering to 8 bpc.
#define ADL_DL_DISPLAY_DITHER_FM8                   3
/// Temporal dithering to 10 bpc.
#define ADL_DL_DISPLAY_DITHER_FM10                  4
/// Spatial dithering to 6 bpc. Note that if the input is 12 bits, the two least significant bits will be truncated.
#define ADL_DL_DISPLAY_DITHER_DITH6                 5

 [... and many more ...]
``` 

 ADL_DL_DISPLAY_DITHER_UNKNOWN is probably not used by ATIRadeonPerf anyway, but seemed to work like ADL_DL_DISPLAY_DITHER_DRIVER_DEFAULT, namely setting the dither state to whatever the driver default is. With all the cards/drivers I tested, dithering was always enabled by default.
 Obviously more problematic is ADL_DL_DISPLAY_DITHER_DISABLED, which is also used by LoadIdentityClut(). However, with the new API it is interpreted as ADL_DL_DISPLAY_DITHER_DRIVER_DEFAULT, so whenever LoadIdentityClut() is trying to disable dithering it actually just selects the drivers default, which most likely is "dithering activated". On the other hand, when trying to enable dithering, the specific dithering, ATIRadeonPerf uses "2", which is ADL_DL_DISPLAY_DITHER_FM6 ("dithering to 6 bpc") and makes ATIRadeonPerf most likely fail because we usually use frame buffers with 8bpc or more, for which "dithering to 6 bpc" is not a valid mode.
 
Even if ADL_Display_DitherState_Set() is used correctly for disabling dithering, still the correct LUT has to be programmed in order to get a dither-fee output **and** to have all pixel values mapped 1:1. Whether such a LUT exists and how it looks like depends on the graphics card and the driver. Interestingly, using the wrong LUT not only leads to missing or duplicated codes, but also can cause dithering, despite having disabled dithering via ADL_Display_DitherState_Set(). On the other hand, disabling dithering is usually required in order to find a dither-free LUT. 
 
 For example, for the HD7750 with Catalyst 13.1WHQL (Jan-2013), for which ATIRadeonPerf still works, none of the more regular LUT formulas seem to do the trick, but a very specific LUT can be found. The same card with the Crimson 15.11WHQL driver (Nov-2015) requires the new API for disabling dithering, and one working LUT is (i<<8) (in C-style as uint16, i=0..255). I also checked the LoadIdentityClut() formulas for gfxhwtype==0, gfxhwtype==1, and gfxhwtype==3 for Windows, but none worked. This was the same with other drivers I tested (>15.11, still HD7750). For the RX460 I had to use the (i<<8) LUT but with the last value set to 0xffff (tested with Crimson 16.9.2 (May'16) and 17.2.1 (Feb'2017)), which turned out to work also for the HD7750 (15.11, 16.9.2, 17.2.1). The same LUT also worked for the R7-260X, but only with Crimson 15.11; I could not find a dither-free LUT with a 1:1 mapping for the RX-260X with Crimson 16.9.2 and 17.2.1 though.

**Comment by kleinerm** [comment]
The values in LoadIdentityClut have been found by painful trial and error by myself and a few others many years ago. Testing happened originally with a CRS Bits+ - visually checking that T-Lock codes got properly recognized and high bit-depths test images don't show dithering artifacts and other artifacts.

Since a couple of years, testing happens with a Datapixx, with the same visual test patterns and automated tests (based on the Datapixx and CRS Bits# ability to capture a scanline of video data and send it back to the host via USB) implemented in DatapixxGPUDitherpatternTest and BitsPlusIdentityClutTest .

For CRS Bits# and the products of VPixx like Datapixx/ViewPixx/Propixx PTB has automated algorithms to find a identity clut via BitsPlusIdentityClutTest and the automated pixel readback from those devices, if none of the cluts in LoadIdentityClut works, implemented in PsychGPUTestAndTweakGammaTables.m
What kind of device do you use if you need identity pixel passthrough? Obviously it is not a CRS or VPixx device?

Testing for proper dither disable and identity gamma tables for untampered pixel passthrough happens frequently on Linux with a collection of graphics cards, for almost each kernel release, way less frequently with a much more limited set of cards on OSX and especially on Windows. E.g., atm. i only have access to one Windows-10 machine with a 8 year old NVidia GeForce 9600 card and a 8 year old AMD Radeon HD-5770. The Radeon is only supported by Catalyst versions up to 15.201.1151.1008 from November 2015, where RadeonPerf still works, so obviously i don't have any modern enough hardware to test more modern Catalyst versions. On that machine RadeonPerf still works perfectly, so i never noticed anything wrong during testing.

This ADL bug you found of changing the meaning/interpretation of hard-coded #define constants is so breathtakingly incompetent, one could almost think one is dealing with an Apple product. Also the SDK versions / Driver interfaces are apparently not versioned, so i don't see how we could work around this. It's basically deciding if we want to break old AMD graphics cards/drivers, or new ones.

Indeed the RadeonPerf code would map "dither off" to "driver default" and "dither on" to 6 bpc temporal dithering (which usually would work on a 8 bpc panel and framebuffer, with the potential artifacts of simulating 8 bpc with 6 bpc output, instead of actually outputting 8 bpc). PTB never calls "dither on" by itself, so effectively this causes a "driver default" setting of likely "no change" when it should disable dithering. That's not great but still allows the user to manually disable dithering in some GUI, i assume?

Linux is not affected by this bug, as we have better methods to disable dithering and guarantee identity gamma tables. Calling RadeonPerf for Linux is just a fallback in case the primary method would fail - something that never happened in 10 years. Also a fallback for the old proprietary Catalyst driver, which is generally not used or needed on Linux with PTB. On Linux + AMD i have ways to usually get such bugs fixed quickly if they would happen (...knocks on wood that it stays that way...).

That using the wrong LUT causes dithering on Windows is something that shouldn't happen if you manually disabled dithering, so you probably should file a bug report to AMD if you can find some way to actually do that for Windows, somewhere where anybody cares. A "wrong" LUT which isn't a strict 8 bit identity LUT should only cause dithering if dithering is enabled in the driver.

Obviously you can send me a pull request for LoadIdentityClut.m with new LUT's and MS-Windows specific detection logic to select the right LUT if you can find a way.

However, see my post to Denis Pelli wrt. gamma table problems on OSX:
https://groups.yahoo.com/neo/groups/PSYCHTOOLBOX/conversations/messages/21754
Issue 3 explained there will likely also increasingly applies to modern AMD graphics cards on Windows, which will make it even more difficult in the future to find suitable identity LUT's for pixel passthrough and may very well defeat all our current workarounds and automated LUT fixing methods on both Windows an d OSX.

**Comment by qx1147** [comment]
First of all, I apologize for responding so late.

> What kind of device do you use if you need identity pixel passthrough? Obviously it is not a CRS or VPixx device?

I don't even really "need" identity pixel passthrough, because I am mainly interested in high temporal resolution rather than contrast or color and I use fast gamer LCDs which usually come with color-wise bad TN panels anyway (6bpc native etc.). Nevertheless, I want to avoid unnecessary dithering as there is enough shimmering on the screen already because of the two dither-stages in the monitor.


> The Radeon is only supported by Catalyst versions up to 15.201.1151.1008 from November 2015, where RadeonPerf still works, so obviously i don't have any modern enough hardware to test more modern Catalyst versions. On that machine RadeonPerf still works perfectly, so i never noticed anything wrong during testing.

So that must have been one of the very last Catalyst drivers before Crimson came out. I cannot tell with which driver the ADL library with the new API was shipped, but I would assume it was the first Crimson version.


> Also the SDK versions / Driver interfaces are apparently not versioned, so i don't see how we could work around this. It's basically deciding if we want to break old AMD graphics cards/drivers, or new ones.

ADL SDK v10 has the functions ADL_Graphics_Versions_Get() and ADL2_Graphics_VersionsX2_Get(), but that doesn't help if these are not supported by earlier versions (ADK v9 has none of them).


> Indeed the RadeonPerf code would map "dither off" to "driver default" and "dither on" to 6 bpc temporal dithering (which usually would work on a 8 bpc panel and framebuffer, with the potential artifacts of simulating 8 bpc with 6 bpc output, instead of actually outputting 8 bpc). 

I am not sure with which framebuffer configuration the mode 2 ("temporal dither to 6 bpc") is supposed to work and whether PTB would ever use such framebuffer configuration. If I use it from a command window on a standard Windows desktop, it fails. This would give us a handle for detecting which ADL API is installed and which value would have to be used for disabling dithering. If not this mode value, probably a different mode value could be found which reliably works with one API but doesn't with the other.

> PTB  never calls "dither on" by itself, so effectively this causes a "driver default" setting of likely "no change" when it should disable dithering. That's not great but still allows the user to manually disable dithering in some GUI, i assume?

No, I never saw a dither-related option.  Maybe one could get away with some registry settings, but this is very messy.


> That using the wrong LUT causes dithering on Windows is something that shouldn't happen if you manually disabled dithering, so you probably should file a bug report to AMD if you can find some way to actually do that for Windows, somewhere where anybody cares. A "wrong" LUT which isn't a strict 8 bit identity LUT should only cause dithering if dithering is enabled in the driver.

Yes, it shouldn't happen and I contacted support already quite a while ago with the expected outcome: Nada!


> However, see my post to Denis Pelli wrt. gamma table problems on OSX:
https://groups.yahoo.com/neo/groups/PSYCHTOOLBOX/conversations/messages/21754
Issue 3 explained there will likely also increasingly applies to modern AMD graphics cards on Windows, which will make it even more difficult in the future to find suitable identity LUT's for pixel passthrough and may very well defeat all our current workarounds and automated LUT fixing methods on both Windows an d OSX.

I am not sure if I ran into the problem you mentioned there. I haven't explained this in my original post, but when my program looks for a dither-free gamma table it checks one 8bit gamma table per screenshot for dither-free table entries, starting with table (i<<8) (= 256 according patches on the screen) and then cycling through all LSBs (256 screenshots). Then it assembles the final gamma table from those entries which resulted in dither-free patches with the correct pixel value. Unfortunately, although the program could find dither-free values for each pixel value, there still was dithering when the finally assembled gamma table was tested, with all modern AMD graphics cards and drivers. I found a working formula only after having looked where the standard formulas actually failed. If I recall correctly, when I looked at the results for (i<<8) I noticed that (255<<8) was mapped to 254 (or at least not to a dither-free 255) and wondered how far I would have to go for a dither-free 255: it was seemingly sufficient to set any bit(s) in the lower byte - with the very surprising side-effect that also all other entries "snapped into place". So there is a clear interaction between the elements which would speak for what you explained in the response to Denis, i.e., that the gamma table is implemented by some kind of interpolation.


> Obviously you can send me a pull request for LoadIdentityClut.m with new LUT's and MS-Windows specific detection logic to select the right LUT if you can find a way.

I can do that, but it makes little sense without having ATIRadeonRef updated. Do you have the source code? In the end, one should know in LoadIdentityClut.m which API had to be used. If it was the new API, one can assume it is Crimson and, if it is Windows, use a new gfxhwtype=5 with my formula. This will not work for all graphics cards and drivers but it will (likely/hopefully) also not brake things that otherwise would have worked. 

Whether the API check is done directly by ATIRadeonRef or in PsychGPUControl.m does not matter, but ATIRadeonRef has to be able to pass mode=0 to the ADL function, no matter which ADL version ATIRadeonRef  is compiled with. PsychGPUControl.m would need a 2nd return parameter to tell LoadIdentityClut.m which API it was. The API test would be assuming the old API and checking whether the function call for enabling dithering failed.

I should add that I always assume that the behavior is the same for all ports (HDMI, DisplayPort, DVI (single or dual)) and across all refresh rates etc., but I can only test with DVI (single link). DisplayPort I don't trust so much in general, and the HDMI handling might also be implemented differently because of the possibly/optionally reduced pixel value range (16-240 instead of 0-255).

**Comment by kleinerm** [comment]
On 04/27/2017 07:08 PM, qx1147 wrote:
> First of all, I apologize for responding so late.
>
>     What kind of device do you use if you need identity pixel
>     passthrough? Obviously it is not a CRS or VPixx device?
>
> I don't even really "need" identity pixel passthrough, because I am
> mainly interested in high temporal resolution rather than contrast or
> color and I use fast gamer LCDs which usually come with color-wise bad
> TN panels anyway (6bpc native etc.). Nevertheless, I want to avoid
> unnecessary dithering as there is enough shimmering on the screen
> already because of the two dither-stages in the monitor.
>
>     The Radeon is only supported by Catalyst versions up to
>     15.201.1151.1008 from November 2015, where RadeonPerf still works,
>     so obviously i don't have any modern enough hardware to test more
>     modern Catalyst versions. On that machine RadeonPerf still works
>     perfectly, so i never noticed anything wrong during testing.
>
> So that must have been one of the very last Catalyst drivers before
> Crimson came out. I cannot tell with which driver the ADL library with
> the new API was shipped, but I would assume it was the first Crimson
> version.
>
>     Also the SDK versions / Driver interfaces are apparently not
>     versioned, so i don't see how we could work around this. It's
>     basically deciding if we want to break old AMD graphics
>     cards/drivers, or new ones.
>
> ADL SDK v10 has the functions ADL_Graphics_Versions_Get() and
> ADL2_Graphics_VersionsX2_Get(), but that doesn't help if these are not
> supported by earlier versions (ADK v9 has none of them).
>

The absence of such functions could maybe be used to detect ADL version 
as well. All functions are dynamically bound at runtime, so a NULL entry 
point at least tells it is v9 on earlier.

>     Indeed the RadeonPerf code would map "dither off" to "driver
>     default" and "dither on" to 6 bpc temporal dithering (which usually
>     would work on a 8 bpc panel and framebuffer, with the potential
>     artifacts of simulating 8 bpc with 6 bpc output, instead of actually
>     outputting 8 bpc).
>
> I am not sure with which framebuffer configuration the mode 2 ("temporal
> dither to 6 bpc") is supposed to work and whether PTB would ever use
> such framebuffer configuration. If I use it from a command window on a
> standard Windows desktop, it fails. This would give us a handle for
> detecting which ADL API is installed and which value would have to be
> used for disabling dithering. If not this mode value, probably a
> different mode value could be found which reliably works with one API
> but doesn't with the other.
>
>     PTB never calls "dither on" by itself, so effectively this causes a
>     "driver default" setting of likely "no change" when it should
>     disable dithering. That's not great but still allows the user to
>     manually disable dithering in some GUI, i assume?
>
> No, I never saw a dither-related option. Maybe one could get away with
> some registry settings, but this is very messy.
>
>     That using the wrong LUT causes dithering on Windows is something
>     that shouldn't happen if you manually disabled dithering, so you
>     probably should file a bug report to AMD if you can find some way to
>     actually do that for Windows, somewhere where anybody cares. A
>     "wrong" LUT which isn't a strict 8 bit identity LUT should only
>     cause dithering if dithering is enabled in the driver.
>
> Yes, it shouldn't happen and I contacted support already quite a while
> ago with the expected outcome: Nada!
>

The classic situation of dealing with a proprietary operating system and 
software when you are not a VIP customer which brings in huge $$$, but 
only small fish. Especially bad when you are a small fish with exotic 
needs > 1-2 stddevs away from the norm, like Neuroscientists are in many 
cases. Hence my push to open-source operating systems whenever possible. 
Works much better there in my experience of the last 10 years. I hope it 
will stay that way.

>     However, see my post to Denis Pelli wrt. gamma table problems on OSX:
>     https://groups.yahoo.com/neo/groups/PSYCHTOOLBOX/conversations/messages/21754
>     Issue 3 explained there will likely also increasingly applies to
>     modern AMD graphics cards on Windows, which will make it even more
>     difficult in the future to find suitable identity LUT's for pixel
>     passthrough and may very well defeat all our current workarounds and
>     automated LUT fixing methods on both Windows an d OSX.
>
> I am not sure if I ran into the problem you mentioned there. I haven't
> explained this in my original post, but when my program looks for a
> dither-free gamma table it checks one 8bit gamma table per screenshot
> for dither-free table entries, starting with table (i<<8) (= 256
> according patches on the screen) and then cycling through all LSBs (256
> screenshots). Then it assembles the final gamma table from those entries
> which resulted in dither-free patches with the correct pixel value.
> Unfortunately, although the program could find dither-free values for
> each pixel value, there still was dithering when the finally assembled
> gamma table was tested, with all modern AMD graphics cards and drivers.
> I found a working formula only after having looked where the standard
> formulas actually failed. If I recall correctly, when I looked at the
> results for (i<<8) I noticed that (255<<8) was mapped to 254 (or at
> least not to a dither-free 255) and wondered how far I would have to go
> for a dither-free 255: it was seemingly sufficient to set any bit(s) in
> the lower byte - with the very surprising side-effect that also all
> other entries "snapped into place". So there is a clear interaction
> between the elements which would speak for what you explained in the
> response to Denis, i.e., that the gamma table is implemented by some
> kind of interpolation.
>

Yes, that sounds like the non-linear lut's are active.

>     Obviously you can send me a pull request for LoadIdentityClut.m with
>     new LUT's and MS-Windows specific detection logic to select the
>     right LUT if you can find a way.
>
> I can do that, but it makes little sense without having ATIRadeonRef
> updated. Do you have the source code? In the end, one should know in
> LoadIdentityClut.m which API had to be used. If it was the new API, one
> can assume it is Crimson and, if it is Windows, use a new gfxhwtype=5
> with my formula. This will not work for all graphics cards and drivers
> but it will (likely/hopefully) also not brake things that otherwise
> would have worked.
>

If you give me your e-mail address (contact me via my gmail address) i 
can mail you a zip file with the source code. The license of the ADL 
SDK, at least the version i used, does not allow distribution of source 
code for applications made with the SDK, only distribution of 
executables. That's why the source code is so far only on my local disc 
drive and must not be publically distributed.

> Whether the API check is done directly by ATIRadeonRef or in
> PsychGPUControl.m does not matter, but ATIRadeonRef has to be able to
> pass mode=0 to the ADL function, no matter which ADL version
> ATIRadeonRef is compiled with. PsychGPUControl.m would need a 2nd return
> parameter to tell LoadIdentityClut.m which API it was. The API test
> would be assuming the old API and checking whether the function call for
> enabling dithering failed.
>
> I should add that I always assume that the behavior is the same for all
> ports (HDMI, DisplayPort, DVI (single or dual)) and across all refresh
> rates etc., but I can only test with DVI (single link). DisplayPort I
> don't trust so much in general, and the HDMI handling might also be
> implemented differently because of the possibly/optionally reduced pixel
> value range (16-240 instead of 0-255).
>

That's not guaranteed at all. The gamma tables will generally behave the 
same, at least for a specific framebuffer color depth (8 bpc vs. > 8 
bpc), but any color space conversions before/after may not, neither will 
dithering.

-mario

**Comment by aproce** [comment]
Slightly unrelated to the topic, I had one query regarding loading identity LUT (disabling LUT) and disabling unnecessary dither in Windows 10 on AMD HD8970M graphics card. I use a Direct3D11 output based DirectShow Media Player which I use for color critical video work. I am on the latest Crimson driver (17.6.2) and was wondering if it is possible to get untouched/bitperfect output from the AMD card. Basically I want that the Desktop Compositor of Windows and the AMD GPU driver should not do any processing, modification or addition of unnecessary dithering to the output of the media player. Can you please suggest what AMD driver settings, registry settings etc. should I use so that the output of the media player is untouched/bitperfect as received by the display?

**Comment by kleinerm** [comment]
Sorry, but that's completely off-topic, your request not even being 
related to Psychtoolbox-3, which uses OpenGL, not Direct3D 11. I can't 
help you. Iff we manage to fix this for Psychtoolbox-3, then maybe you 
can use the improved ATIRadeonperf_Windows.exe to disable dithering, but 
other than that, there ain't nothing to recommend here.

**Comment by kleinerm** [comment]
Closing this issue, as it is resolved in current PTB 3.0.14 betas by Marcs contribution.

**Comment by qx1147** [comment]
Oh my, this seems to be a moving target. The finding so far was that LUT#0 (LUTtype=0 in LoadIdentityClut.m) would not work for the graphics cards I tested, so LUT#5 was implemented and is currently taken as default for AMD cards. However, according to my most recent tests with the AMD graphics card drivers Adrenalin 2019 (19.9.2) and Adrenalin 2020 (20.1.3) (with Win10-EDU-1803), LUT#0 is good and LUT#5 is bad - in most cases anyway.

The good news, I don't get any unexpected dithering anymore, so if I don't measure the identity table, it is not because of dithering but because the pixel value mapping is not 1:1. Often, just the upper part of the table (>240) would be shifted by one; or almost all the table would be shifted by one except for the first 8 entries, meaning that the table is actually okay for the most (relevant) part.

Anyway, here are my test results. "OK" means I measured the identity table, "x" means there was a mapping error, "-" means I was too lazy for making the measurements. Unless otherwise mentioned, the results were the same for the DisplayPort (DP) and the DVI outputs. Using DVI via the DisplayPort output and a passive DP2DVI converter cable always follows the DisplayPort behavior.
```
                  19.9.2/LUT#0   19.9.2/LUT#5   20.1.3/LUT#0   20.1.3/LUT#5
RadeonPRO WX5100      x              OK              x             OK
Radeon R7-250X        OK             OK              OK            OK
Radeon HD7750         OK             OK              -             -       
Radeon R7-260X        OK             x               OK            x
Radeon RX550 (DP)     OK             x               OK            x
    -"-      (DVI)    OK             OK              OK            x
Radeon RX460 (DP)     OK             x               -             -
    -"-      (DVI)    OK             OK              -             -
```
Note that the RadeonPRO WX5100 runs with either the normal Adrenalin driver (which automatically switches into a PRO mode) or with the "Radeon Pro Software Enterprise" driver, and the results regarding the LUTs are the same. However, the Enterprise driver does not allow LoadIdentityClut to turn on/off dithering. LoadIdentityClut detects this problem, assumes that the card or driver must be old, and chooses LUT#0 rather than LUT#5 as the default, which is exactly the wrong one in this case.

Even though things seem to be still in a move, LUT#0 seems to be obviously the better default with newer drivers. Unfortunately, I don't know when this changed. I don't think it is a Win10 vs. Win7 issue, because I observed the old behavior (LUT#0 bad, LUT#5 good) under both Win7 and Win10. Whatever, I guess the advise for the user would be anyway as usual: if you want to be sure, measure it yourself for your own specific setup.

I use a video capture device (Epiphan DVI2USB 3.0) and a couple of splitters and signal converters for making these measurements. The capture device allows me to check for dithering, at least when the signal is just 8bpc. Moreover, the measurement itself is relative quick; setting it up is what takes time, especially when the capture device doesn't like the video timing that is fed to the monitor, or vice versa. But since dithering seems to be no issue anymore, one could verify the LUTs also with luminance measurements, which wouldn't require such specific hardware and fiddly setup.
